### ISSUE TITLE
Grid

### ISSUE THREAD

**Issue by vijayholimath** [bespoke-silicon-group/bsg_replicant#519]
I have one fundamental question. In the Github repo:  https://github.com/bespoke-silicon-group/bsg_replicant/blob/master/regression/cuda/test_float_all_ops.c

Line no 50-64, the function  host_float_all_ops(...) is called as expected result at line no 391 but the actual results are coming from res_add_host, res_sub_host, res_mul_host, res_div_host, etc. May I ask where do we set-up host fpu operation for add, sub, mul, div, etc. since  host_float_all_ops(...) is the expected result? I am unable to figure-out in the source code perhaps I am missing something.

**Comment by drichmond** [comment]
I believe @bornaehsani wrote this test. He should be able to answer your question

**Comment by vijayholimath** [comment]
@bornaehsani I guess to set-up FPU operation for host FPU  one can set in the directory bsg_manycore/software/spmd/bsg_cuda_lite_runtime/ with respective applications For example  test_float_all_ops.c in the directory  bsg_replicant/regression/cuda/
 gets host FPU result from bsg_manycore/software/spmd/bsg_cuda_lite_runtime/float_all_ops/ ? Please correct if I am wrong, Makefile in bsg_manycore/software/spmd/bsg_cuda_lite_runtime/float_all_ops/ is written for VCS co-simulation not for F1 execution?  If it is written for F1 execution, please let me know how to run Make.

Many Thanks,
-Vijay

**Comment by save-buffer** [comment]
This test is making sure that the FPU on the manycore is producing correct results. We define "correct" results as those matching some other FPU that we trust is correct. In this case, we trust the FPU on the host processor. 
So this test: 
Allocates input and output arrays on the device (i.e. the manycore), copies input values, runs the kernel defined in `bsg_manycore/software/spmd/bsg_cuda_lite_runtime/float_all_ops/kernel_float_all_ops.c` on the manycore, then copies the results back to the host. The host then performs the same operations on its own (in `host_float_all_ops`) and then compares the results. The host FPU does not require any setup. 
We do expect some small error between the two, but as long as the error is within a threshold we define, the test will pass. 

In regards to running on F1, after mounting the AGFI for the manycore, you should be able to go into `bsg_replicant/regression/cuda` and run `sudo make test_float_all_ops.log`, and it will run the test.

**Comment by bornaehsani** [comment]
Hi @vijayholimath . So as @save-buffer has explained, the `host_float_all_ops()` is purely to make sure the results we get back from the device are correct. As the purpose of the regression tests are to in fact verify the correctness of the Manycore operations, we perform the same computation on both host (in this case `host_float_all_ops()` ) and the device (the actual kernel that is run), and then compare the results. As to your question regarding setting up the FPU operations in the host, I don't completely understand what you mean. There is no setup necessary for running FPU operations on the host, the entire host code located inside `bsg_bladerunner/bsg_replicant/regression/cuda` runs on the X86 host, including the `host_float_all_ops()`. The only setup necessary is for the device code. Does that answer your question?

**Comment by vijayholimath** [comment]
Thanks for the clarification save-buffer and bornaehsani. Actually, I want to run only certain operations of my choice, let's say, three operations, add, mul and float to int. How do I run make on F1? I did comment out certain operations which are not of interest in both directory source codes:

bsg_replicant/regression/cuda/test_float_all_ops.c line no. 56- 61

bsg_manycore/software/spmd/bsg_cuda_lite_runtime/float_all_ops/kernel_float_all_ops.c line 27-39

In bsg_replicant/regression/cuda/tests.mk I commented out all tests except  test_float_all_ops (line no 68)

then I ran "make regression" in bsg_replicant/regression/cuda/ directory after loading AGFI, I am still seeing results from host FPU operations which are not interest to me.  I am using BSG repo version 3.5.2.

By the way, we have our own RISC-V compiler and we want to compile the codes using our compiler. I think the file to point one's own compiler is at:  bsg_manycore/software/mk/Makefile.builddefs line no. 58-63?

I think default tile is set to 2X2 where do I set tile numbers? I think maximum one can tile is up to 256X256 since the HammerBlade supports up to 512 cores?

Many Thanks,
-Vijay

**Comment by save-buffer** [comment]
The results of the operations are printed [in this loop](https://github.com/bespoke-silicon-group/bsg_replicant/blob/dde5081a2a61cc56942772045692cf07d189296e/regression/cuda/test_float_all_ops.c#L427). You can skip operations that do not interest you. Alternatively, you could restructure the code by removing the operations from the `operations` array that you don't need, remove them from both `host_float_all_ops` and `kernel_float_all_ops`, and remove the allocations and copying of the input/output arrays. 

`bsg_manycore/software/mk/Makefile.builddefs` lines 58-63 are correct for modifying the compiler you're using. 

What do you mean by "setting tile numbers"? If you mean changing the dimensions of the manycore array, I believe you would have to recompile the manycore and create a new AGFI. I'm not sure how to do that, that's @drichmond's area of expertise. If you're interested in remapping the tile indices in a tilegroup to different physical tiles than the default, this is not supported. Also, `256x256` would be a 65536-tile manycore array. I'm not sure an F1 FPGA could hold that many.

**Comment by vijayholimath** [comment]
Thanks save-buffer..:) Sorry I meant changing the dimensions of many core array. Dustin, can you please tell me where to change tile dimension of many core array? I plan to change different configurations such as 10X10, 20X20, etc. I think it is some where in *.v code as a parameter and re-compile the code and generate AGFI to new dimensions. Apart from that, I think one has to change some where in software to initialise the devices?

**Comment by vijayholimath** [comment]
Hi @drichmond,  I think many array dimensions can be changed at: bsg_manycore/syn/Makefile.dc from line  no. 52-53? And one may need to change other parameters as well from line no. 47-54? How about in software, do we need to change some where for memory initialisation and others?

Thanks,
-Vijay

**Comment by bornaehsani** [comment]
Hi Vijay, Dustin is on vacation, I think @vb000 has experience with building the AGFI with desired parameters. Bandhav do you know how to build custom AGFI? Thanks!

**Comment by vb000** [comment]
Hi @vijayholimath, I had built agfi quite a while back per Dustin's instructions, now concisely captured in `bsg_bladerunner`'s [`RELEASE.md`](https://github.com/bespoke-silicon-group/bsg_bladerunner/blob/master/RELEASE.md).

That looks up-to-date, could you try it out and let us know specific questions, if any?

**Comment by vijayholimath** [comment]
Hello @vb000, I can sucessfully build AFI and AMI, I don't have issue with that. My question is rather, where to change manycore dimension array and rebuild AFI? I think now the array is set to 2X2, I would like to change to different configurations such as 10X10, 15X15, etc

**Comment by vb000** [comment]
All machine configuration is sourced from a file called `Makefile.machine.include` and that's here:
https://github.com/bespoke-silicon-group/bsg_replicant/blob/master/Makefile.machine.include

Could you try changing here?

**Comment by vijayholimath** [comment]
Hello @save-buffer or @bornaehsani or any FP experts, I sucessfully replaced my arithmetic units with FP add, FP mul, int to float and float to int in fpu_float.v and fpu_int.v. Now, all cuda floating point test works fine such as test_float_vec_*.c where * here is add, mul, div, sqrt, etc. except log and exp is not working. If I try to run for test_float_vec_exp and test_float_vec_log, regression get stuck. Since div, sqrt, log, exp, etc. are FP subroutines (written in assembly). I am not getting how to debug exp and log functions, any suggestions?

**Comment by vb000** [comment]
Does your toolchain build `spike`? If so, you could write an assembly test that runs on a single tile and compare the intermediate results of emulated FP functions in the trace, with those in spike's trace.

**Comment by vijayholimath** [comment]
@vb000, thanks for the file.  But in https://github.com/bespoke-silicon-group/bsg_replicant/blob/master/Makefile.machine.include I don't see any parameter which is set to 2, I mean something like num_x_tiles:= 2,  num_y_tiles:= 2, which matches with 2X2 tile description. Perhaps I am missing something? Or you mean by default this is set to 4X5, line no. 28 and 29?

**Comment by vb000** [comment]
Yes, `BSG_MACHINE_GLOBAL_X` and `BSG_MACHINE_GLOBAL_Y` are what might be interested in... default is 4x4 machine (`BSG_MACHINE_GLOBAL_Y` also includes IO on the top row of the grid).

**Comment by vijayholimath** [comment]
Yes @vb000  the toolchain does support 'spike'. Unfortunately, I don't know much about spike and how to write test for spike and compare results. Any guidance much appreciated...:)

**Comment by vb000** [comment]
Spike is RISC-V ISA simulator which you could you to simulate RISC-V executables. If you have RISC-V executable like `prog.riscv`, it can be executed on spike by `spike prog.riscv`. You could find a usage tutorial here: https://github.com/riscv/riscv-isa-sim.

__A crazy thought for FPU debugging:__

Is it possible to add some non-synthesizable logic to your FPU that checks the result result of your FPU with that of x86. May be something like:

```
logic [31:0] my_result;
my_fpu_mul(.a_i(a), .b_i(b), .y_o(my_result));

// synopsys translate_off
assert (my_result == a * b);
// synopsys translate_on
```

My experience has been that our FPU's result should match x86 result exactly (bitwise), unless we make a conscious decision to deviate from the FP standard slightly. In which case, you could modify the assert accordingly.

**Comment by vijayholimath** [comment]
Thanks @vb000. I did add some logic to our arithmetic circuits and debugged using ila probes and checked on the board in cloud with that of x86 FPU. In our case, results deviate from x86 FPU and this is expected because we have different number system other than IEEE-754 floating point number system.  

I was checking BSG repo, spike is defined in https://github.com/bespoke-silicon-group/bsg_manycore/blob/master/software/mk/Makefile.builddefs in line no. 63 and each cuda test cases such as for exp in https://github.com/bespoke-silicon-group/bsg_manycore/blob/master/software/spmd/bsg_cuda_lite_runtime/float_vec_exp/Makefile  in line no. 30 generates it's own main.riscv  But I haven't  found it's usage in BSG repo. Perhaps I might have overlooked? If you could tell me usage or executing for example or add in BSG repo float_vec_exp test case that will be great?  Actually, I am looking for instruction by instruction of exp function in FP subroutine so that it is easier to point-out where test is getting stuck..:)

Many thanks.

**Comment by vijayholimath** [comment]
I think BSG HammerBlade don't  have opcodes for FP sqrt, div, exp and log because these functions are not coded in BSG_FPU rather they are soft FP subroutine. In other words, these functions are taken care by RISC-V C/C++ compiler as a fplib.so I think. So I think using 'Spike' I don't think it is useful for instruction level debugging?

**Comment by vb000** [comment]
> I think BSG HammerBlade don't have opcodes for FP sqrt, div, exp and log because these functions are not coded in BSG_FPU rather they are soft FP subroutine. In other words, these functions are taken care by RISC-V C/C++ compiler as a fplib.so I think. So I think using 'Spike' I don't think it is useful for instruction level debugging?

Compile the spike binary with the same options you use for compiling the binary for manycore, I think it's `-mno-fdiv` and more... this makes sure compiler use emulated instructions.

**Comment by vijayholimath** [comment]
Sorry I am not clear. Yes in BSG repo mno-fdiv swith is off as the makefile line no. 32 in: https://github.com/bespoke-silicon-group/bsg_manycore/blob/master/software/riscv-tools/Makefile

Since fdiv, fexp, etc. are off these instructions are emulated. But my question was how to debug for emulated instructions? I am not clear in spike, as you know spike requires opcodes, but in RISC-V instruction I haven't found instruction and opcode for fexp and flog under 'RV32F Standard Extension' as per the manual page 106 in: https://content.riscv.org/wp-content/uploads/2017/05/riscv-spec-v2.2.pdf

**Comment by dpetrisko** [comment]
Hi, in order to add a custom instruction (flog, fexp) to spike, take a look at https://github.com/riscv/riscv-isa-sim/blob/master/README.md#simulating-a-new-instruction. 

To compile a program with the new instructions (without modifying the compiler toolchain), you can use inline assembly in the program to emit raw bytes which correspond to your custom encoding. To add fdiv/fsqrt instructions simply remove the corresponding flags from the compilation command in the Makefile.

**Comment by vijayholimath** [comment]
> you can use inline assembly in the program to emit raw bytes which correspond to your custom >encoding. To add fdiv/fsqrt instructions simply remove the corresponding flags from the >compilation command in the Makefile.

It would be great if you could send me example code and Makefile for fexp?

**Comment by vb000** [comment]
Use the same binary with both spike and manycore. Then, if an instruction is emulated for manycore, it'll be emulated for spike as well.

Comparing traces can be quite a bit of work as manycore and spike binaries won't match exactly (the part of emulated instructions would match exactly though). Consider writing assertions in verilog that does fp bitwise matching for you.

**Comment by drichmond** [comment]
> All machine configuration is sourced from a file called `Makefile.machine.include` and that's here:
> https://github.com/bespoke-silicon-group/bsg_replicant/blob/master/Makefile.machine.include
> 
> Could you try changing here?

This is the correct method for changing the AGFI/AFI dimensions.

**Comment by vijayholimath** [comment]
I was working on spike, I think it would make sense if you add new instruction but I think fexp is not an instruction rather it is a fp subroutine algorithm. In other words, whenever fexp is found in assembly code/source code, the compiler would  then call fexp algorithm not an fexp instruction?

**Comment by vijayholimath** [comment]
BTW, in https://github.com/bespoke-silicon-group/bsg_replicant/blob/master/Makefile.machine.include line 29 (BSG_MACHINE_GLOBAL_Y                  = 5) why it is 5? I was under impression that AGFI/AFI dimensions always in terms of square, for example 4X4, 5X5, 6X6, etc. Perhaps I misunderstood something?

**Comment by drichmond** [comment]
Because that's the size of entire machine, including IO and memory routers. 

If you want the dimensions of the manycore array: 

CL_MANYCORE_DIM_Y                    := $(shell expr $(BSG_MACHINE_GLOBAL_Y) - 1)

**Comment by vijayholimath** [comment]
OK, got it. But whenever I run cuda regression test (make regression) in bsg_replicant/regression/cuda/ I get BSG_INFO test is running on 2X2 tile group but I think  currently it is set to 4X5 as per: https://github.com/bespoke-silicon-group/bsg_replicant/blob/master/Makefile.machine.include
Could you please help me to understand?

**Comment by save-buffer** [comment]
There is a distinction between the size of the manycore and a tilegroup: The size of the manycore is the number of physical cores. A tilegroup is an abstraction on top of that that gives your kernel the illusion that it's running on its own manycore array. Several tilegroups can execute in parallel. For example, if you have a 4x4 manycore, you could run four kernels that run in a 2x2 tilegroup at the same time. 

If you are familiar with CUDA terminology, one tile in a tilegroup is analogous to a thread and a tilegroup is analogous to a block.

**Comment by vijayholimath** [comment]
Thank you very much @save-buffer ..:)

**Comment by vijayholimath** [comment]
@vb000 or any FP experts.  I was working on spike, I think it would make sense if you add new instruction but I think fexp is not an instruction rather it is a fp subroutine algorithm. In other words, whenever fexp is found in assembly code/source code, the compiler would then call fexp algorithm not an fexp instruction?

**Comment by vijayholimath** [comment]
Since exponential and logarithmic functions are  fp subroutines in bsg RISC-V compiler tool chain, I have to accept it is not possible to debug fp subroutines. Unless, if we add custom instruction and modify compiler to support custom instruction.  Please correct me if I am wrong?

**Comment by vb000** [comment]
> Use the same binary with both spike and manycore. Then, if an instruction is emulated for manycore, it'll be emulated for spike as well.
> 
> Comparing traces can be quite a bit of work as manycore and spike binaries won't match exactly (the part of emulated instructions would match exactly though). Consider writing assertions in verilog that does fp bitwise matching for you.

^^^ You could debug emulated instructions (fp subroutines in toolchain) this way.

**Comment by vijayholimath** [comment]
Thanks @vb000..:) I will try. Off the the FPU Operation thread, I have a question regarding AMI. After creating AMI and launching AMI instance, I am not clear about usage of the launched AMI instance. I think any c/cpp applications can  be written using BSG API  without including concerned header files because they are hidden in AMI kernel? It will be great if you could educate me on usage of AMI instance.

**Comment by vijayholimath** [comment]
Just wondering, at what frequency bladerunner running on F1 instance?

**Comment by drichmond** [comment]
125 MHz - the default.

**Comment by drichmond** [comment]
We haven't spent any time optimizing it so that is not a maximum frequency.

**Comment by vijayholimath** [comment]
Got it. Would bladerunner architecture suitable for mobile devices?

**Comment by vijayholimath** [comment]
Hello All,

I just recompiled design in AWS with no modification, it taking more than 9 hours  synthesise and implement the design. After 9 hours it is still in synthesis step!  About couple of months ago, the same design took only 5 hours to synthesise and implement the design.  I am unable to figure-out what's happening, why recompiling taking longer time. Any suggestions?

Many Thanks.

**Comment by drichmond** [comment]
Hi @vijayholimath 

I don't have any specific suggestions. We haven't made any major changes to the hardware in the last couple of months, and last time I compiled it took a few hours (not overnight/9 hours).

**Comment by vijayholimath** [comment]
I solved the problem...:) The issue is, if I use latest version of FPGA developer AMI: 1.8.0, it takes very long time to synthesize but with FPGA developer AMI: 1.7.0 it takes only about 5 hours. This is strange, I think in the latest version the design might be optimising for frequency. By the way, by any chance, do you have ASIC implementation report for latest technology node say, 7nm or 10nm or any other nodes?

**Comment by vijayholimath** [comment]
Hello All,

I did very minute changes in /bsg_bladerunner/bsg_manycore/software/spmd/bsg_cuda_lite_runtime/scalar_print/  and
/bsg_bladerunner/bsg_manycore/software/spmd/fib I am trying to regenerate main.riscv, bsg_set_tile_x_y.o and bsg_printf.o in respective directories  by running 'make regression' in  bsg_f1/regression/spmd and running 'sudo make install'  and 'make build' in bsg_f1/libraries. Unfortunately, I am unable to regenerate these files, I mean these files are not updating. Can you please let me know how to update these files? I am using FPGA Image Version: 3.5.2

Many Thanks,
-Vijay

**Comment by vijayholimath** [comment]
Does BSG manycore/F1 support OpenCL?

**Comment by save-buffer** [comment]
No, it currently only supports only our CUDA Lite programming model.

**Comment by vijayholimath** [comment]
Hmmm, I think supporting OpenCL is not so complicated. I would very much like to support it, unfortunately it is very much architecture dependent and I don't have much knowledge on BSG many core architecture. I think your team can finish OpenCL project in a week since you people are experts....:)

**Comment by mrutt92** [comment]
It's not a priority for us. It sounds like it might be for you. Aren't you interested in learning more about our architecture?

**Comment by vijayholimath** [comment]
I am learning BSG bladerunner architecture and it is quite interesting architecture. Unfortunately, I do not have much human resource and it will take lot of time in my case...:)

**Comment by taylor-bsg** [comment]
Thanks for your suggestion Vijay! Maybe one day somebody will contribute a
port!

**Comment by vijayholimath** [comment]
Sure, Michael...:) Many Thanks!

**Comment by vijayholimath** [comment]
Hello All,

Can you please tell me how you generated riscv32-unknown-elf-dramfs-c++ , riscv32-unknown-elf-dramfs-ar, riscv32-unknown-elf-dramfs-gcc , etc. which are in bsg_manycore/software/riscv-tools/riscv-install/bin?  I think you used github repo which is mentioned in bsg_manycore/software/riscv-tools/Makefile as a NEWLIB_URL and in some other makefiles as well  as a cloned github repo. But other than this there is no directory named "bsg_newlib_dramfs" in the project.  The reason why I am asking these details because we have our own compiler we want to generate our own riscv32-unknown-elf-dramfs-c++ , riscv32-unknown-elf-dramfs-ar, riscv32-unknown-elf-dramfs-gcc , etc. 

Many Thanks,
-Vijay

**Comment by dpetrisko** [comment]
Hi Vijay —

bsg_newlib_dramfs is substituted for vanilla newlib here in the makefiles. 

https://github.com/bespoke-silicon-group/bsg_manycore/blob/master/software/riscv-tools/Makefile#L41

https://github.com/bespoke-silicon-group/bsg_manycore/blob/master/software/riscv-tools/newlib.patch

It comes from our fork of newlib found here:
https://github.com/bespoke-silicon-group/bsg_newlib_dramfs

The cross-compile targets are set up to use this tuple, so you’ll have to replace it with your own:
https://github.com/bespoke-silicon-group/bsg_manycore/blob/master/software/mk/Makefile.builddefs#L59

**Comment by vijayholimath** [comment]
Thank you very much Dustin..:) I know above line numbers in Makefiles very well. But what I am unable to figure is, how bsg_newlib_dramfs is associated with building compiler. I use cross tool ng (https://crosstool-ng.github.io/) to build our compiler. Can you please tell me how I do associate https://github.com/bespoke-silicon-group/bsg_newlib_dramfs with crosstool-ng?

**Comment by vijayholimath** [comment]
I think in crosstool-ng we need to apply this patch: https://github.com/bespoke-silicon-group/bsg_manycore/blob/master/software/riscv-tools/newlib.patch while building compiler. Unfortunately, I don't know how to apply this patch. Any assistance highly appreciated. If you wish I can share config file for building in crosstool-ng? Many Thanks!

**Comment by vijayholimath** [comment]
I am not sure, using this Makefile: https://github.com/bespoke-silicon-group/bsg_manycore/blob/master/software/riscv-tools/Makefile one can download/obtain source codes for compilers and rebuild compiler after applying patches in source codes?

**Comment by vijayholimath** [comment]
Hello @dpetrisko , @drichmond  and the team just to inform you, I solved the patching newlib.path to our compiler...:)  We need to build gfortran compliler as well, in $(RISCV_BIN_DIR), I did not find gfortran compiler may be this is because BSG project do not require this. 

I think in order to build fortran compiler one need to enable switch  --enable--languages =c,c++,fortran. May I ask in which part of the Makefile your only enabling languages 'only c,c++' so that I can add extra fortran switch? I found this switch is used in lot of places in /riscv-gnu-tool-chain/Makefile, I could manually try enabling each places, since building compiler in our case takes more than 6 hours trying out all possible takes may be take more than 3 days. Since you people are experts, I thought I can ask you exact place to enable the fortran switch..:)  Many Thanks!

**Comment by save-buffer** [comment]
The makefile that builds the toolchain clones the official `riscv-gnu-toolchain` repo and configures and builds that. In [this](https://github.com/riscv/riscv-gnu-toolchain/blob/master/Makefile.in#L532) Makefile rule, you can see that the enabled languages are `C` and `C++`. `stamps/build-gcc-newlib-stage2` is the rule that generates the final RISC-V compiler.

**Comment by vijayholimath** [comment]
Thank you very much @save-buffer ...:)

**Comment by vijayholimath** [comment]
@vb000 or the team, I want to use only one RISC-V core for testing purpose, I guess I need to change BSG_MACHINE_GLOBAL_X =1 and BSG_MACHINE_GLOBAL_Y =1 in bsg_f1/Makefile.machine.include? I guess there are no settings for software? I am using v3.5.2 bsg repo

Thanks,
-Vijay

**Comment by vijayholimath** [comment]
In bsg_f1/Makefile.machine.include,  I have changed BSG_MACHINE_GLOBAL_X =1 and BSG_MACHINE_GLOBAL_Y =2,  to use 'only one core', but synthesis getting failed. With following error:

ERROR: [Synth 8-524] part-select [30:6] out of range of prefix 'addr' [..../bsg_f1/build/src_post_encryption/bsg_cache_to_axi_hashed.v:189]

I think this error is related to BSG_MACHINE_DRAM_SIZE_WORDS = 536870912, the size of this DRAM is 2^29.  Can you please educate me how's the mathematical relations (for BSG_MACHINE_DRAM_SIZE_WORDS)  for different BSG_MACHINE_GLOBAL_X,  BSG_MACHINE_GLOBAL_Y sizes for example: 4*4, 2*2, 1*1, 3*3...n*n?

Many Thanks!

**Comment by dpetrisko** [comment]
Known issue: https://github.com/bespoke-silicon-group/bsg_manycore/issues/315

**Comment by vijayholimath** [comment]
I changed BSG_MACHINE_GLOBAL_X =3 and BSG_MACHINE_GLOBAL_Y =2 but still synthesis is failing. Perhaps I am missing something?

**Comment by vijayholimath** [comment]
I want to compile all programs in bsg_f1/regression/cuda and bsg_manycore/software/spmd/bsg_cuda_lite_runtime/float_vec_* to C++. I think I need to convert all files to *.cpp and then compile with CXX flag. I want to change flag CC to CXX somewhere in Makefiles. Can you please tell me where are those files to change the flag? I tried with grep command, I am getting lot of hits and I am running out of screen...:)
